### PR TITLE
flock a file before running pg_dump to avoid stacking scheduled runs

### LINKED ISSUE #34: Long execution and cron scheduling can lead to processes stacking up
As mentioned in #33, a locked relation can block the execution pg_dump forever. We could end up with many pg_back processes stacking up if the script is schedule by cron or a similar tool.

A simple solution would be implement a per-database locking mecanism to avoid concurrent runs of pg_dump in this context.

### PR DESCRIPTION
**Pull request by orgrim**
Closes #34

### CODE REVIEW

File: pg_back [hunk 1]
```diff
@@ -297,6 +297,25 @@ do
 	continue
     fi
 
+    # Try to lock a file named after to database we are going to dump
```
Comment (pgstef) [inline pg_back]: "named after THE database" ?

**Comment by pgstef** [comment]
Global test (running 2 pg_back in //):

```
2020-03-27 11:44:55 CET  INFO: preparing to dump
2020-03-27 11:44:55 CET  INFO: target directory is /var/lib/pgsql/12/backups/
2020-03-27 11:44:55 CET  INFO: listing databases
2020-03-27 11:44:55 CET  INFO: dumping global objects into /var/lib/pgsql/12/backups//pg_global_2020-03-27_11-44-55.sql
2020-03-27 11:44:55 CET  INFO: saving output of SHOW ALL to /var/lib/pgsql/12/backups//pg_settings_2020-03-27_11-44-55.out
2020-03-27 11:44:55 CET  INFO: acquiring internal lock for postgres
2020-03-27 11:44:55 CET  INFO: dumping database "postgres" into /var/lib/pgsql/12/backups//postgres_2020-03-27_11-44-55.dump
2020-03-27 11:44:56 CET  INFO: acquiring internal lock for db1
2020-03-27 11:44:56 CET  INFO: dumping database "db1" into /var/lib/pgsql/12/backups//db1_2020-03-27_11-44-56.dump
2020-03-27 11:44:57 CET  INFO: acquiring internal lock for db2
2020-03-27 11:44:57 CET  INFO: dumping database "db2" into /var/lib/pgsql/12/backups//db2_2020-03-27_11-44-57.dump
2020-03-27 11:45:16 CET  INFO: acquiring internal lock for db3
2020-03-27 11:45:16 CET  WARNING: unable to obtain the lock on db3, skipping this database
2020-03-27 11:45:16 CET  INFO: acquiring internal lock for db4
2020-03-27 11:45:16 CET  INFO: dumping database "db4" into /var/lib/pgsql/12/backups//db4_2020-03-27_11-45-16.dump
2020-03-27 11:45:42 CET  INFO: acquiring internal lock for db5
2020-03-27 11:45:42 CET  WARNING: unable to obtain the lock on db5, skipping this database
2020-03-27 11:45:42 CET  INFO: purging old backups
2020-03-27 11:45:44 CET  INFO: done
```

```
2020-03-27 11:45:05 CET  INFO: preparing to dump
2020-03-27 11:45:05 CET  INFO: target directory is /var/lib/pgsql/12/backups
2020-03-27 11:45:05 CET  INFO: listing databases
2020-03-27 11:45:05 CET  INFO: dumping global objects into /var/lib/pgsql/12/backups/pg_global_2020-03-27_11-45-05.sql
2020-03-27 11:45:05 CET  INFO: saving output of SHOW ALL to /var/lib/pgsql/12/backups/pg_settings_2020-03-27_11-45-05.out
2020-03-27 11:45:05 CET  INFO: acquiring internal lock for postgres
2020-03-27 11:45:05 CET  INFO: dumping database "postgres" into /var/lib/pgsql/12/backups/postgres_2020-03-27_11-45-05.dump
2020-03-27 11:45:06 CET  INFO: acquiring internal lock for db1
2020-03-27 11:45:06 CET  INFO: dumping database "db1" into /var/lib/pgsql/12/backups/db1_2020-03-27_11-45-06.dump
2020-03-27 11:45:09 CET  INFO: acquiring internal lock for db2
2020-03-27 11:45:09 CET  WARNING: unable to obtain the lock on db2, skipping this database
2020-03-27 11:45:09 CET  INFO: acquiring internal lock for db3
2020-03-27 11:45:09 CET  INFO: dumping database "db3" into /var/lib/pgsql/12/backups/db3_2020-03-27_11-45-09.dump
2020-03-27 11:45:35 CET  INFO: acquiring internal lock for db4
2020-03-27 11:45:35 CET  WARNING: unable to obtain the lock on db4, skipping this database
2020-03-27 11:45:35 CET  INFO: acquiring internal lock for db5
2020-03-27 11:45:35 CET  INFO: dumping database "db5" into /var/lib/pgsql/12/backups/db5_2020-03-27_11-45-35.dump
2020-03-27 11:45:52 CET  INFO: purging old backups
2020-03-27 11:45:52 CET  INFO: done
```

Lock file is detected and pg_dump for that database is skipped. That's interesting.
However, that would mean, when the 1st pg_dump will be released, the original pg_back will continue, and dump the other databases. We'll end up with potentially multiple dumps of the same database. We can probably live with that limitation though.

Also, if pg_back is (i. e.) killed, the lock file stays. It should be mentioned somewhere to check for that .lock file? Wouldn't it be a bit dangerous to activate that behaviour by default then?

**Comment by orgrim** [comment]
Hello,

> Lock file is detected and pg_dump for that database is skipped. That's interesting.
> However, that would mean, when the 1st pg_dump will be released, the original pg_back will continue, and dump the other databases. We'll end up with potentially multiple dumps of the same database. We can probably live with that limitation though.

Each run shall dump as many db as required or possible. Commit c8895d418d8bd63f97c93ef60a304cd0efae4b1f ensures the date in the filename is a close as the pg_dump execution as possible.
 
> Also, if pg_back is (i. e.) killed, the lock file stays. It should be mentioned somewhere to check for that .lock file? Wouldn't it be a bit dangerous to activate that behaviour by default then?

The file stays but it is no longer locked, the next run just reuse it. One may use fuser or lslocks to check if the lockfile is not stale.

Regards,

**Comment by orgrim** [comment]
Thanks for the contribution!

**Pull request merged**